Firestore/Source/Core/FSTFirestoreClient.mm

```diff
@@ -118,6 +123,9 @@ - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
       }
     };
 
+    _credentialsProvider->SetUserChangeListener(
+        [userChangeListener](const User &user) { userChangeListener(user); });
+
```

Comment:
Could you change lines 111-128 to read like this:
```
_credentialsProvider->SetUserChangeListener( [whateverYouNeedHere](const User& user) {
  FSTStrongify(self);
  ...
});
```

i.e. move the objc block into the c++ lambda (thus eliminating the objc block entirely)

Caveat: I have *no* idea how objc blocks and c++ lambdas interact, so this may be a bad idea.

Optional. (Especially if this is a bad idea. :)

Reply:
This should work especially because there's no interaction between blocks and lambdas here.

Reply:
Okay, so the problem here is that `__block` variables are an Objective-C feature that lifts the variable into the closure of the block. The calling stack frame (that creates the block) no longer has a copy of the variable and refers to the value by reference.

There's no equivalent feature in C++ lambdas, and `__block` variables can't be captured in a C++ lambda.

Luckily the pair of `(initialUserAvailable, initialUser)` was a manual implementation of a future using a `dispatch_semaphore`. Converting this code to an `std::future` has helped simplify it.